### PR TITLE
cp: cleanup serviceprofiles in core and extensions

### LINKED ISSUE #6084: Service Profiles for core cp and extensions
Currently, There are [service profile resources available](https://github.com/linkerd/linkerd2/blob/main/cli/installsp/template.go#L4) for the core
control-plane in the `diagnostics install-sp` command through which they can be
installed on-demand.

These are currently out-dated, and may not work as expected with all the recent
changes that the control-plane components went through.

One of the reason for having these (I think) into a separate `install-sp` cmd instead of
`linkerd install` is that having both the `ServicProfile` definiation and instances
in the same manifest could cause problem with the install, and there are long standing
issues around this in Helm, etc.

With the recent addition of extensions, This won't be the problem as CRD is part of
the core install, while extension components can get Service Profiles instances as
part of their install.

## Proposed Changes

With the above infromation in mind, I propose the following changes:

- Clean up `diagnostics install-sp` with the latest core cp changes.
- All extension components should have their service profile definiations
  as part of their install. (atleast first party extensions i.e viz, multi-cluster, etc)
  (Should they be enabled by default, or manually enabled is a point to be discussed)
- Optionally, Extensions should also get their own `install-sp` cmd to install
  service profile definations later on.
  (Implementation wise, Both the above points can be simplified, by having good code re-use)

@olix0r

### PR DESCRIPTION
**Pull request by Pothulapati**
Fixes #6084 

Because of all the changes in control-plane and extensions, The
current serviceprofiles in `install-sp` are not relevant anymore.

Most of the current ones, should be moved over into the viz extension.

This leaves the core cp with just the `linkerd-dst` serviceprofile
(which can maybe removed)

Signed-off-by: Tarun Pothulapati <tarunpothulapati@outlook.com>

### CODE REVIEW

File: viz/charts/linkerd-viz/templates/service-profiles.yaml [hunk 1]
```diff
@@ -21,52 +22,24 @@ spec:
...
-      pathRegex: /io\.linkerd\.proxy\.destination\.Destination/GetProfile
+      pathRegex: /api/v1/Edges
 ---
 apiVersion: linkerd.io/v1alpha2
 kind: ServiceProfile
 metadata:
-  name: linkerd-prometheus.linkerd.svc.cluster.local
-  namespace: linkerd
+  name: prometheus.{{.Values.namespace}}.svc.{{.Values.clusterDomain}}
+  namespace: {{.Values.namespace}}
```
Comment (alpeb) [inline viz/charts/linkerd-viz/templates/service-profiles.yaml:42]: I think the prometheus and grafana parts should be guarded around those components `.enabled` condition

**Review by alpeb — COMMENTED** [review]
What do you think about taking this opportunity to add the missing routes for tap, tap-injector and web? :wink:

**Comment by alpeb** [comment]
Also, the `routes_test.go` integration test will need to be updated given the absence of the Destination routes

File: viz/cmd/install.go [hunk 2]
```diff
@@ -36,6 +36,7 @@ var (
 		"templates/tap-injector-rbac.yaml",
 		"templates/tap-injector.yaml",
 		"templates/web.yaml",
+		"templates/service-profiles.yaml",
```
Comment (dadjeibaah) [inline viz/cmd/install.go:39]: Super specific nit: This is unrelated to your change but the slice variable that holds this string is named `templatesVIz`, I think it should be `templateViz` instead.

File: cli/cmd/diagnostics.go [hunk 3]
```diff
@@ -36,7 +36,6 @@ This command provides subcommands to diagnose the functionality of Linkerd.`,
 	diagnosticsCmd.AddCommand(newCmdControllerMetrics())
 	diagnosticsCmd.AddCommand(newCmdEndpoints())
 	diagnosticsCmd.AddCommand(newCmdMetrics())
-	diagnosticsCmd.AddCommand(newCmdInstallSP())
```
Comment (dadjeibaah) [inline cli/cmd/diagnostics.go:39]: Now that we are removing this command, we also need to update the instruction in the help text to remove this part
```
...
  # Install service profiles for the control-plane components.
  linkerd diagnostics install-sp
```

**Review by dadjeibaah — COMMENTED** [review]
Changes are looking great. I left a few comments for your consideration. 

Looks like we need to also update the `bin/test-scale` script to remove the occurences of `linkerd install-sp`
```bash
(⎈ |kind-dev:default)➜  linkerd2 git:(tarun/sp-cp-ext) grep -ir -I "install-sp" --exclude CHANGES.md
bin/test-scale:"$linkerd_path" -l "$linkerd_namespace" install-sp | kubectl apply -f -
cli/cmd/diagnostics.go:  linkerd diagnostics install-sp
```

File: viz/cmd/install.go (hunk 2 above)
Comment (Pothulapati) [reply viz/cmd/install.go:39]: Done!

File: viz/charts/linkerd-viz/templates/service-profiles.yaml (hunk 1 above)
Comment (Pothulapati) [reply viz/charts/linkerd-viz/templates/service-profiles.yaml:43]: Makes sense. Updated!

**Comment by alpeb** [comment]
Sorry for being such a nag, but wdyt about replacing the first part of the routes integration test that was pointing to the linkerd namespace and now it only returns [DEFAULT] routes, with one hitting the linkerd-viz namespace, so it can check against all its routes?
```
 bin/linkerd viz routes -n linkerd-viz deploy
==> deployment/grafana <==
ROUTE                                             SERVICE   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
GET /api/annotations                              grafana         -     -             -             -             -
GET /api/dashboard/{dashboard}                    grafana         -     -             -             -             -
GET /api/dashboards/tags                          grafana         -     -             -             -             -
GET /api/dashboards/uid/{uid}                     grafana         -     -             -             -             -
GET /api/datasources/proxy/1/api/v1/query_range   grafana         -     -             -             -             -
GET /api/datasources/proxy/1/api/v1/series        grafana         -     -             -             -             -
GET /api/search                                   grafana         -     -             -             -             -
GET /d/{uid}/{dashboard-name}                     grafana         -     -             -             -             -
GET /public/build/{style}.css                     grafana         -     -             -             -             -
GET /public/fonts/{font}                          grafana         -     -             -             -             -
GET /public/img/{img}                             grafana         -     -             -             -             -
[DEFAULT]                                         grafana         -     -             -             -             -

==> deployment/metrics-api <==
ROUTE                           SERVICE   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
POST /api/v1/Edges          metrics-api         -     -             -             -             -
POST /api/v1/Gateways       metrics-api         -     -             -             -             -
POST /api/v1/ListPods       metrics-api         -     -             -             -             -
POST /api/v1/ListServices   metrics-api         -     -             -             -             -
POST /api/v1/SelfCheck      metrics-api         -     -             -             -             -
POST /api/v1/StatSummary    metrics-api         -     -             -             -             -
POST /api/v1/TopRoutes      metrics-api         -     -             -             -             -
[DEFAULT]                   metrics-api         -     -             -             -             -

==> deployment/prometheus <==
ROUTE                        SERVICE   SUCCESS   RPS   LATENCY_P50   LATENCY_P95   LATENCY_P99
GET /api/v1/query_range   prometheus         -     -             -             -             -
GET /api/v1/series        prometheus         -     -             -             -             -
POST /api/v1/query        prometheus         -     -             -             -             -
[DEFAULT]                 prometheus         -     -             -             -             -
```

**Comment by Pothulapati** [comment]
@alpeb Updated the integration test to also check for routes in the viz namespace

**Review by dadjeibaah — APPROVED** [review]
📦 🚢

**Review by alpeb — APPROVED** [review]
Thanks @Pothulapati !

**Comment by Pothulapati** [comment]
@alpeb Before merging this, reg SP's for `web`, `tap-injector` and `tap`. I couldn't find both for `tap-injector` and `tap` as the requests to them are through API Service and somehow `profile --tap` was not able to generate them, It did generate the routes for healthcheck requests and the same happened for other resources too 🤔

File: test/integration/routes/routes_test.go [hunk 4]
```diff
@@ -35,16 +40,10 @@ func TestRoutes(t *testing.T) {
 		testutil.AnnotatedFatal(t, "'linkerd routes' command failed", err)
 	}
 
-	routeStrings := []struct {
-		s string
-		c int
-	}{
+	routeStrings := []testCase{
```
Comment (adleong) [inline test/integration/routes/routes_test.go:43]: TIOLI but I preferred the way this was before with an anonymous struct

**Review by adleong — APPROVED** [review]

**Comment by alpeb** [comment]
> @alpeb Before merging this, reg SP's for `web`, `tap-injector` and `tap`. I couldn't find both for `tap-injector` and `tap` as the requests to them are through API Service and somehow `profile --tap` was not able to generate them, It did generate the routes for healthcheck requests and the same happened for other resources too

Right, the `tap-injector` and `tap` connections from the k8s API are TLS'd so they skip the proxy. OTOH I think we can get something for `web`. But would have to run it using `Ingress`, otherwise `linkerd viz dashboard` port-forwards directly into the container and the proxy is bypassed as well.

**Pull request merged**